extensions/rich_text_components/components.py

```diff
@@ -0,0 +1,172 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Classes for Rich Text Components in Oppia."""
+
+import re
+
+import bs4
+from extensions.objects.models import objects
+
+
+class BaseRteComponent(object):
+    """Base object class.
+
+    This is the superclass for rich text components in Oppia, such as
+    Image and Video.
+    """
+
+    # These values should be overridden in subclasses.
+    customization_args = {}
+
+    arg_schema = {}
+
+    @classmethod
+    def validate(cls, value_dict):
+        """Validates customization args for a rich text component.
+
+        Raises:
+          TypeError: if any customization arg is invalid.
+        """
+        if cls.arg_schema:
+            required_attrs = cls.arg_schema
+        else:
+            required_attrs = cls.customization_args
+        required_attr_names = required_attrs.keys()
+        attr_names = value_dict.keys()
+        if set(attr_names) != set(required_attr_names):
+            raise Exception('Invalid attributes')
+
+        for arg_name in required_attr_names:
+            arg_obj_class = required_attrs[arg_name]
+            arg_obj_class.normalize(value_dict[arg_name])
+
+
+class Collapsible(BaseRteComponent):
+    """Class for Collapsible component."""
+
+    customization_args = {
+        'heading-with-value': objects.UnicodeString,
+        'content-with-value': objects.Html
+    }
+
+    @classmethod
+    def validate(cls, value_dict):
+        """Validates Collapsible component."""
+        super(Collapsible, cls).validate(value_dict)
+        content = value_dict['content-with-value']
+        inner_soup = bs4.BeautifulSoup(
+            content.encode(encoding='utf-8'),
+            'html.parser')
+        collapsible = inner_soup.findAll(
+            name='oppia-noninteractive-collapsible')
+        tabs = inner_soup.findAll(
+            name='oppia-noninteractive-tabs')
+        if len(collapsible) or len(tabs):
+            raise Exception('Nested tabs and collapsible')
+
+
+class Image(BaseRteComponent):
+    """Class for Image component."""
+
+    customization_args = {
+        'filepath-with-value': objects.Filepath,
+        'caption-with-value': objects.UnicodeString,
+        'alt-with-value': objects.UnicodeString
+    }
+
+    @classmethod
+    def validate(cls, value_dict):
+        """Validates Image component."""
+        super(Image, cls).validate(value_dict)
+        filename_re = r'^[A-Za-z0-9+/]*\.((png)|(jpeg)|(gif)|(jpg))$'
+        filepath = value_dict['filepath-with-value']
+        if not re.match(filename_re, filepath):
+            raise Exception('Invalid filepath')
+
+
+class Link(BaseRteComponent):
+    """Class for Link component."""
+
+    customization_args = {
+        'text-with-value': objects.UnicodeString,
+        'url-with-value': objects.SanitizedUrl
+    }
+    is_simple = True
```

Comment:
Why is there no is_simple in the base class and in other components?

Probably best to set it to None in the base class and have a backend test to verify it's a boolean everywhere.

Reply:
`is_simple` is not needed. I had added that earlier but found `is_complex` in component definitions and then removed it. I missed it here, sorry for that.